### PR TITLE
[Do Not Merge] Implementing Device.command as a decorator

### PR DESCRIPTION
**Pull request by bohrium272**
- [ ] Obtain the device with the defined command at L#515
- [ ] Implement tests

### CODE REVIEW

**Comment by bohrium272** [comment]
@mtambos I require some help. As you can see in the code L#515 of `server.py` references `calling_device` to assign a route of the command to the device's `Flask` app 
The issue I'm facing is I haven't found a way to obtain such a reference in the `command` decorator

File: MOCC/src/configurations/Comm-Layer/src/server.py [hunk 1]
```diff
@@ -472,50 +474,48 @@ def __init__(self, fget=None, **kwargs):
...
+        'definitions': {},
+        'responses': {}
+    }
+
+    for arg in args:
+        swagger_dict['parameters'].append({
+            "name": arg,
+            "in": form,
+            "type": obj,
+            "required": false,
```
Comment (mtambos) [inline MOCC/src/configurations/Comm-Layer/src/server.py]: Values in `swagger_dict` don't make sense.

File: MOCC/src/configurations/Comm-Layer/src/server.py [hunk 2]
```diff
@@ -472,50 +474,48 @@ def __init__(self, fget=None, **kwargs):
...
+        })
+
+    for arg, def_value in args_with_default_values:
+        swagger_dict['parameters'].append({
+            "name": arg,
+            "in": form,
+            "type": obj,
+            "required": true,
```
Comment (mtambos) [inline MOCC/src/configurations/Comm-Layer/src/server.py]: Same as previous comment

File: MOCC/src/configurations/Comm-Layer/src/server.py [hunk 3]
```diff
@@ -472,50 +474,48 @@ def __init__(self, fget=None, **kwargs):
...
-            return req.json()
-        return _command_executor
-    return command_wrapper
\ No newline at end of file
+    }
+    command_ip = '/command/' + str(_command.__name__)
+    calling_device.server_app.route(command_ip)(_command)
+    swag_from()(swagger_dict)
+    def _command_executor(*args, **kwargs):
+        req = requests.request('POST', command_ip)
```
Comment (mtambos) [inline MOCC/src/configurations/Comm-Layer/src/server.py]: here you should call `f` with the arguments received through a POST requests.
This is a server, you receive request, and then redirect them to the underlying method that's being decorated with `@command`

File: MOCC/src/configurations/Comm-Layer/src/server.py [hunk 4]
```diff
@@ -472,50 +474,48 @@ def __init__(self, fget=None, **kwargs):
...
+        return req.json()
```
Comment (mtambos) [inline MOCC/src/configurations/Comm-Layer/src/server.py]: The return value should be `flask.jsonify(f(*args, **kwargs))`

File: MOCC/src/configurations/Comm-Layer/src/server.py [hunk 5]
```diff
@@ -465,25 +486,50 @@ def __init__(self, fget=None, **kwargs):
...
+
+    swagger_dict['responses'] = {
+        '200': {
+            'description': inspect.getdoc(_command)
+        }
+    }
+    
+    def _command_executor(*args, **kwargs):
+        arg_dict = request.values.to_dict()
+        return flask.jsonify(f(arg_dict))
```
Comment (mtambos) [inline MOCC/src/configurations/Comm-Layer/src/server.py]: here you should probably use `f(**arg_dict)`

File: MOCC/src/configurations/Comm-Layer/src/tests/test_server.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    def __init__(self):
+        super().__init__()
+
+    @command
+    def test_command(self, arg1, arg2=None):
+        return "999"
+
+
+dp = DeviceProxy()
+dp.test_command(10)
```
Comment (mtambos) [inline MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: this does not test how we want the `@command` decorator to function.
you should test whether the calls to `test_command` received through HTTP are redirected to the implementation.
You could use Flask's test utilities for instance http://flask.pocoo.org/docs/0.12/testing/

**Comment by bohrium272** [comment]
@mtambos I've added Swagger capabilities, however server.py is failing the Unit Test and when the request is sent through Swagger, the result code is 400

**Comment by mtambos** [comment]
Have you trying debugging it?
You could add an error handler to the SeverBase, and put a breakpoint in the handler.
The you can use the traceback module to get a stack trace of the original exception.
Maybe the Swagger definition is not correct, or the REST call does not match the Swagger definition?

http://flask.pocoo.org/docs/0.12/patterns/apierrors/
https://docs.python.org/3/library/traceback.html

**Comment by bohrium272** [comment]
Yes I've found out that the request parameters aren't being passed properly, hence the right side of [this line](https://github.com/mars-planet/mars_city/pull/62/commits/d30bc41b5fa8ece37ace76bebc41f45efc8787ec#diff-4b8a31c3ed7b9339b5f506009e73269cR533) returns `None`

**Comment by mtambos** [comment]
do you want to try to solve the problem, or should I?

**Comment by bohrium272** [comment]
I'll investigate and fix it 😄

**Comment by bohrium272** [comment]
@mtambos I can't seem to find why the request params come out to be `None`, I tried some other techniques too using request parser but couldn't get it to work.

File: MOCC/src/configurations/Comm-Layer/src/tests/test_server.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    yield client
+
+
+def test_cmd1(client):
+    data = {
+        'arg1': str(uuid.uuid4()),
+        'arg2': str(uuid.uuid4()),
+    }
+    rv = client.post('/cmd1')
```
Comment (mtambos) [inline MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: hey Arpit, sorry it took me so long to answer.
Here you forgot to pass `data` to the POST request.
the call should be: `rv = client.post('/cmd1', json=data)`
with that you receive the data in [here](https://github.com/mars-planet/mars_city/pull/62/commits/d30bc41b5fa8ece37ace76bebc41f45efc8787ec#diff-4b8a31c3ed7b9339b5f506009e73269cR533)

Comment (bohrium272) [reply MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: This is the error that I get when I put `json=data`
>E       TypeError: __init__() got an unexpected keyword argument 'json'

But when I change it to `data=data` this is what I get
screen shot 2018-07-21 at 12 18 28 am

Comment (mtambos) [reply MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: Did you check the version of your libraries? Are the stable versions?
What about Python's version?

Comment (mtambos) [reply MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: You shouldn't use the `data` parameter.
If you do, then you need to change the code where you read from `requests.json`

Comment (bohrium272) [reply MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: According to this https://stackoverflow.com/a/46991833/4014694 it's fine to do so

Comment (bohrium272) [reply MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: > If you do, then you need to change the code where you read from requests.json

What change needs to be made? From what I can see the request object has pretty much only the URL

Comment (mtambos) [reply MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: The thread you linked to refers to the `requests` library, but you're sending the request using the test client from Flask.
I don't know exactly what changes you need to make, if I did, I'd just do it myself.
If you feel you cannot continue with this, I can pick up the task from here.

Comment (bohrium272) [reply MOCC/src/configurations/Comm-Layer/src/tests/test_server.py]: That would be beneficial..

**Review by mtambos — APPROVED** [review]
I made everything work, the requirements.txt file has the versions of the libraries used

**Pull request merged**